src/samples/design_patterns/stream_sample.c

```diff
@@ -0,0 +1,285 @@
+{
+    struct my_stream_api_handle *handle = data;
+    ssize_t status;
+
+    //Append data to the buffer
+    status = my_stream_device_read(handle->dev, sol_buffer_at_end(&handle->rx), handle->rx.capacity - handle->rx.used);
+    if (status < 0) {
+        if (status == EAGAIN || status == EINTR)
+            return true;
+        else {
+            SOL_WRN("Could not read to the stream device!");
+            handle->read_monitor = NULL;
+            return false;
+        }
+    }
+
+    handle->rx.used += status;
+    //Keep reading until there's no more data
+    if (!handle->read_timeout)
+        handle->read_timeout = sol_timeout_add(0, _inform_user, handle);
+
+    return true;
+}
+
+//Stream creation, where rx_cb is configured
+struct my_stream_api_handle *
+my_stream_api_new(const struct my_stream_api_config *config, int dev)
+{
+    struct my_stream_api_handle *handle;
+    size_t rx_size;
+    void *buf = NULL;
+    enum sol_buffer_flags flags = SOL_BUFFER_FLAGS_NO_NUL_BYTE;
+
+    handle = calloc(1, sizeof(struct my_stream_api_handle));
+    SOL_NULL_CHECK(handle, NULL);
+
+    handle->tx_cb = config->tx_cb;
+    handle->tx_size = config->tx_size;
+    handle->user_data = config->user_data;
+
+    //The user does not want to read from the stream, ignore rx configuration.
+    if (config->rx_cb) {
+        handle->rx_cb = config->rx_cb;
+        rx_size = config->rx_size;
+
+        if (rx_size) {
+            buf = malloc(rx_size);
+            SOL_NULL_CHECK_GOTO(buf, err_buf);
+        } else
+            flags |= SOL_BUFFER_FLAGS_DEFAULT;
+
+        //Setup input monitor
+        handle->read_monitor = my_stream_device_add_io_monitor(handle->dev, SOL_FD_FLAGS_IN, _can_read, handle);
+        SOL_NULL_CHECK_GOTO(handle->read_monitor, err_monitor);
+        flags |= SOL_BUFFER_FLAGS_FIXED_CAPACITY;
+    } else
+        flags |= SOL_BUFFER_FLAGS_DEFAULT;
+
+    sol_buffer_init_flags(&handle->rx, buf, rx_size, flags);
+    sol_ptr_vector_init(&handle->pending_blobs);
+    handle->dev = dev;
+
+    return handle;
+
+err_monitor:
+    free(buf);
+err_buf:
+    free(handle);
+    return NULL;
+}
+
+void
+my_stream_api_close(struct my_stream_api_handle *handle)
+{
+    uint16_t i;
+    struct sol_blob *blob;
+
+    //The user is trying to delete the handle from the rx_cb, do not delete it now.
+    if (handle->in_use) {
+        handle->delete_me = true;
+        return;
+    }
+
+    SOL_PTR_VECTOR_FOREACH_IDX (&handle->pending_blobs, blob, i)
+        sol_blob_unref(blob);
+    sol_ptr_vector_clear(&handle->pending_blobs);
+    sol_buffer_fini(&handle->rx);
+    if (handle->read_timeout)
+        sol_timeout_del(handle->read_timeout);
+    if (handle->read_monitor)
+        my_stream_device_remove_io_monitor(handle->read_monitor);
+    if (handle->write_monitor)
+        my_stream_device_remove_io_monitor(handle->write_monitor);
+    free(handle);
+}
+
+//! [stream read]
+
+static void
+startup(void)
+{
```

Comment:
you should open it here, so it's run-able. You can use dev = STDIN_FILENO or STDOUT_FILENO